# Segmenting and Clustering Neighborhoods in Toronto City

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [2]:
from bs4 import BeautifulSoup # this module helps in web scrapping.


### Dataset

In [3]:
url= "https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&direction=prev&oldid=926287641"

In [4]:
data = requests.get(url).text

In [5]:
soup= BeautifulSoup(data, "html5lib")

In [6]:
tables= soup.find_all('table')

In [7]:
len(tables)

5

In [8]:
tables

[<table class="wikitable sortable">
 <tbody><tr>
 <th>Postcode</th>
 <th>Borough</th>
 <th>Neighbourhood
 </th></tr>
 <tr>
 <td>M1A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M2A</td>
 <td>Not assigned</td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M3A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
 </td></tr>
 <tr>
 <td>M4A</td>
 <td><a href="/wiki/North_York" title="North York">North York</a></td>
 <td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
 </td></tr>
 <tr>
 <td>M5A</td>
 <td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
 <td><a class="mw-redirect" href="/wiki/Harbourfront_(Toronto)" title="Harbourfront (Toronto)">Harbourfront</a>
 </td></tr>
 <tr>
 <td>M5A</td>
 <td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
 <td><a href="/wiki/Regent_Park" title="Re

In [9]:
for index,table in enumerate(tables):
    if ("Postcode" in str(table)):
        table_index = index
print(table_index)

0


Analyzing from above two code cells, It appears that the table index of the table we need is '0'.

In [10]:
print(tables[table_index].prettify())

<table class="wikitable sortable">
 <tbody>
  <tr>
   <th>
    Postcode
   </th>
   <th>
    Borough
   </th>
   <th>
    Neighbourhood
   </th>
  </tr>
  <tr>
   <td>
    M1A
   </td>
   <td>
    Not assigned
   </td>
   <td>
    Not assigned
   </td>
  </tr>
  <tr>
   <td>
    M2A
   </td>
   <td>
    Not assigned
   </td>
   <td>
    Not assigned
   </td>
  </tr>
  <tr>
   <td>
    M3A
   </td>
   <td>
    <a href="/wiki/North_York" title="North York">
     North York
    </a>
   </td>
   <td>
    <a href="/wiki/Parkwoods" title="Parkwoods">
     Parkwoods
    </a>
   </td>
  </tr>
  <tr>
   <td>
    M4A
   </td>
   <td>
    <a href="/wiki/North_York" title="North York">
     North York
    </a>
   </td>
   <td>
    <a href="/wiki/Victoria_Village" title="Victoria Village">
     Victoria Village
    </a>
   </td>
  </tr>
  <tr>
   <td>
    M5A
   </td>
   <td>
    <a href="/wiki/Downtown_Toronto" title="Downtown Toronto">
     Downtown Toronto
    </a>
   </td>
   <td>
    <a class=

Now, we will convert above table into pandas dataframe.

In [11]:
df = pd.DataFrame(columns=["Postcode", "Borough", "Neighbourhood"])

for row in tables[table_index].tbody.find_all("tr"):
    col = row.find_all("td")
    if (col != []):
        postcode = col[0].text.strip()
        
        borough = col[1].text.strip()
        neighbourhood = col[2].text.strip()
        if borough!= "Not assigned" :
            
            df = df.append({"Postcode":postcode, "Borough":borough, "Neighbourhood":neighbourhood}, ignore_index=True)


# print df

df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
30       M9C         Etobicoke   
31       M9C         Etobicoke   
32       M9C         Etobicoke   
33       M1E       Scarborough   
34       M1E       Scarborough   
35       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
46       M3H        North York   
47       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
50       M5H  Downtown Toronto   
51       M5H  Downtown Toronto   
52       M6H      West Toronto   
53       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
56       M2J        North York   
57       M2J        North York   
58       M3J        North York   
59       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
62       M5J  Downtown Toronto   
63       M5J  Downtown Toronto   
64       M6J      West Toronto   
65       M6J      West Toronto   
66       M1K       Scarborough   
67       M1K       Scarborough   
68       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
71       M3K        North York   
72       M4K      East Toronto   
73       M4K      East Toronto   
74       M5K  Downtown Toronto   
75       M5K  Downtown Toronto   
76       M6K      West Toronto   
77       M6K      West Toronto   
78       M6K      West Toronto   
79       M1L       Scarborough   
80       M1L       Scarborough   
81       M1L       Scarborough   
82       M2L        North York   
83       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
86       M4L      East Toronto   
87       M5L  Downtown Toronto   
88       M5L  Downtown Toronto   
89       M6L        North York   
90       M6L        North York   
91       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
94       M1M       Scarborough   
95       M1M       Scarborough   
96       M2M        North York   
97       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
101      M5M        North York   
102      M6M              York   
103      M6M              York   
104      M6M              York   
105      M6M              York   
106      M9M        North York   
107      M9M        North York   
108      M1N       Scarborough   
109      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
115      M6N              York   
116      M9N              

In [12]:
df.loc[16,'Neighbourhood']= "Cloverdale Mall"
df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M6A        North York   
5        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
9        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
12       M4B         East York   
13       M5B  Downtown Toronto   
14       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
17       M9B         Etobicoke   
18       M9B         Etobicoke   
19       M9B         Etobicoke   
20       M9B         Etobicoke   
21       M1C       Scarborough   
22       M1C       Scarborough   
23       M1C       Scarborough   
24       M3C        North York   
25       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
30       M9C         Etobicoke   
31       M9C         Etobicoke   
32       M9C         Etobicoke   
33       M1E       Scarborough   
34       M1E       Scarborough   
35       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
46       M3H        North York   
47       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
50       M5H  Downtown Toronto   
51       M5H  Downtown Toronto   
52       M6H      West Toronto   
53       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
56       M2J        North York   
57       M2J        North York   
58       M3J        North York   
59       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
62       M5J  Downtown Toronto   
63       M5J  Downtown Toronto   
64       M6J      West Toronto   
65       M6J      West Toronto   
66       M1K       Scarborough   
67       M1K       Scarborough   
68       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
71       M3K        North York   
72       M4K      East Toronto   
73       M4K      East Toronto   
74       M5K  Downtown Toronto   
75       M5K  Downtown Toronto   
76       M6K      West Toronto   
77       M6K      West Toronto   
78       M6K      West Toronto   
79       M1L       Scarborough   
80       M1L       Scarborough   
81       M1L       Scarborough   
82       M2L        North York   
83       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
86       M4L      East Toronto   
87       M5L  Downtown Toronto   
88       M5L  Downtown Toronto   
89       M6L        North York   
90       M6L        North York   
91       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
94       M1M       Scarborough   
95       M1M       Scarborough   
96       M2M        North York   
97       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
101      M5M        North York   
102      M6M              York   
103      M6M              York   
104      M6M              York   
105      M6M              York   
106      M9M        North York   
107      M9M        North York   
108      M1N       Scarborough   
109      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
115      M6N              York   
116      M9N              

More than one neighborhood can exist in one postal code area. So, We will combine rows with same postcode into one row by seperating other neighbourhood name with comma.

In [13]:
drop_row_list=[]
for i in range(len(df)):
    
    for j in range(i+1, len(df)):
        if df.loc[i,'Postcode']==df.loc[j,"Postcode"] and df.loc[i, "Borough"]==df.loc[j, "Borough"]:
            df.loc[i,'Neighbourhood']= df.loc[i,'Neighbourhood']+ ', ' + df.loc[j, "Neighbourhood"]
            drop_row_list.append(j)
drop_row_list= set(drop_row_list)

x=list(drop_row_list)


now, drop the duplicate rows with same postcode.

In [14]:
df.drop(x, axis=0, inplace=True)
df

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
52       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
58       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
64       M6J      West Toronto   
66       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
72       M4K      East Toronto   
74       M5K  Downtown Toronto   
76       M6K      West Toronto   
79       M1L       Scarborough   
82       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
87       M5L  Downtown Toronto   
89       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
96       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
102      M6M              York   
106      M9M        North York   
108      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
116      M9N              York   
117      M1P       Scarborough   
120      M2P        North York   
121      M4P   Central Toronto   
122      M5P   Central Toronto   
124      M6P      West Toronto   
126      M9P         Etobicoke   
127      M1R       Scarborough   
129      M2R        North York   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
4                     Lawrence Heights, Lawrence Manor  
6                                         Not assigned  
7                                     Islington Avenue  
8                                       Rouge, Malvern  
10         

Now, if the Neigbourhood name is 'Not assigned', we will replace it with its borough name.

In [15]:
for  i,row in df.iterrows():
    if df.loc[i, 'Neighbourhood']=='Not assigned':
        df.loc[i, 'Neighbourhood']= df.loc[i, 'Borough']
df        
    
    
    
            
            

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
52       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
58       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
64       M6J      West Toronto   
66       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
72       M4K      East Toronto   
74       M5K  Downtown Toronto   
76       M6K      West Toronto   
79       M1L       Scarborough   
82       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
87       M5L  Downtown Toronto   
89       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
96       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
102      M6M              York   
106      M9M        North York   
108      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
116      M9N              York   
117      M1P       Scarborough   
120      M2P        North York   
121      M4P   Central Toronto   
122      M5P   Central Toronto   
124      M6P      West Toronto   
126      M9P         Etobicoke   
127      M1R       Scarborough   
129      M2R        North York   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
4                     Lawrence Heights, Lawrence Manor  
6                                         Queen's Park  
7                                     Islington Avenue  
8                                       Rouge, Malvern  
10         

df.reset_index(drop=True) #reset the index.

In [17]:
df.shape

(103, 3)

So, now the dataframe has 103 rows and 3 columns.

### Finding Latitude and Longitude

In [18]:
df_postcode = df.groupby(['Postcode', 'Borough'])
df['Neighbourhood'].apply(', '.join).reset_index()

df_postcode.head()

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
52       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
58       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
64       M6J      West Toronto   
66       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
72       M4K      East Toronto   
74       M5K  Downtown Toronto   
76       M6K      West Toronto   
79       M1L       Scarborough   
82       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
87       M5L  Downtown Toronto   
89       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
96       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
102      M6M              York   
106      M9M        North York   
108      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
116      M9N              York   
117      M1P       Scarborough   
120      M2P        North York   
121      M4P   Central Toronto   
122      M5P   Central Toronto   
124      M6P      West Toronto   
126      M9P         Etobicoke   
127      M1R       Scarborough   
129      M2R        North York   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Harbourfront, Regent Park  
4                     Lawrence Heights, Lawrence Manor  
6                                         Queen's Park  
7                                     Islington Avenue  
8                                       Rouge, Malvern  
10         

Now, we will get the latitude and longitude values from the csv file given in the url below.

In [19]:
locgeo_df = pd.read_csv('https://cocl.us/Geospatial_data', index_col='Postal Code')



In [46]:
toronto_data = df.join(locgeo_df, on='Postcode')
toronto_data

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
52       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
58       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
64       M6J      West Toronto   
66       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
72       M4K      East Toronto   
74       M5K  Downtown Toronto   
76       M6K      West Toronto   
79       M1L       Scarborough   
82       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
87       M5L  Downtown Toronto   
89       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
96       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
102      M6M              York   
106      M9M        North York   
108      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
116      M9N              York   
117      M1P       Scarborough   
120      M2P        North York   
121      M4P   Central Toronto   
122      M5P   Central Toronto   
124      M6P      West Toronto   
126      M9P         Etobicoke   
127      M1R       Scarborough   
129      M2R        North York   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Harbourfront, Regent Park  43.654260 -79.360636  
4                     Lawrence Heights, Lawrence Manor  43.718518 -79.464763  
6                                         Queen's Park  43.662301 -79.38

### Exploring the Neighbourhoods of Toronto.

#### Create a map of Toronto with neighborhoods superimposed on top.


In [21]:
address = 'Toronto'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto city are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto city are 43.6534817, -79.3839347.


In [22]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Define Foursquare Credentials and Version


In [23]:
CLIENT_ID = 'GXPUANRGBMKVVADOO51YPDAITVHXNEWAF213X5J32UVQHSY5' # your Foursquare ID
CLIENT_SECRET = 'SGMWBY0VMKA0Z0CD2WUKNT2IMCSCHD4LCUT4A3YZR4RINHHG' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GXPUANRGBMKVVADOO51YPDAITVHXNEWAF213X5J32UVQHSY5
CLIENT_SECRET:SGMWBY0VMKA0Z0CD2WUKNT2IMCSCHD4LCUT4A3YZR4RINHHG


#### Now, let's get the nearby venues that are in Toronto within a radius of 500 meters.


In [24]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called toronto_venues_.


In [25]:

toronto_venues= getNearbyVenues(names=toronto_data['Neighbourhood'],
                                  latitudes= toronto_data['Latitude'],
                                  longitudes= toronto_data['Longitude'])

Parkwoods
Victoria Village
Harbourfront, Regent Park
Lawrence Heights, Lawrence Manor
Queen's Park
Islington Avenue
Rouge, Malvern
Don Mills North
Woodbine Gardens, Parkview Hill
Ryerson, Garden District
Glencairn
Cloverdale Mall, Islington, Martin Grove, Princess Gardens, West Deane Park
Highland Creek, Rouge Hill, Port Union
Flemingdon Park, Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Bloordale Gardens, Eringate, Markland Wood, Old Burnhamthorpe
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Downsview North, Wilson Heights
Thorncliffe Park
Adelaide, King, Richmond
Dovercourt Village, Dufferin
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto
Harbourfront East, Toronto Islands, Union Station
Little Portugal, Trinity
East Birchmount Park, Ionview, Kennedy Park
Bayview Village
CFB Toronto, Downsview East


In [26]:
print(toronto_venues.shape)
toronto_venues.head()

(2125, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2         Parkwoods               43.753259               -79.329656   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1                     KFC       43.754387       -79.333021   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1   Fast Food Restaurant  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

Let's check how many venues were returned for each neighborhood


In [27]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                                96   
Agincourt                                                                4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                       3   
Albion Gardens, Beaumond Heights, Humbergate, J...                       9   
Alderwood, Long Branch                                                   7   
Bathurst Manor, Downsview North, Wilson Heights                         19   
Bayview Village                                                          4   
Bedford Park, Lawrence Manor East                                       25   
Berczy Park                                                             58   
Birch Cliff, Cliffside West                                              4   
Bloordale Gardens, Eringate, Markland Wood, Old...                       9   
Brockton, Exhibition Place, Parkdale Village                            22   
Business Reply Mail Processing Centre 969 Eastern                       16   
CFB Toronto, Downsview East                                              2   
CN Tower, Bathurst Quay, Island airport, Harbou...                      17   
Cabbagetown, St. James Town                                             43   
Caledonia-Fairbanks                                                      4   
Canada Post Gateway Processing Centre                                   13   
Cedarbrae                                                                8   
Central Bay Street                                                      62   
Chinatown, Grange Park, Kensington Market                               60   
Christie                                                                16   
Church and Wellesley                                                    77   
Clairlea, Golden Mile, Oakridge                                          9   
Clarks Corners, Sullivan, Tam O'Shanter                                 12   
Cliffcrest, Cliffside, Scarborough Village West                          2   
Cloverdale Mall, Islington, Martin Grove, Princ...                       2   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              36   
Davisville North                                                         9   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                      14   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                           5   
Design Exchange, Toronto Dominion Centre                               100   
Don Mills North                                                          5   
Dorset Park, Scarborough Town Centre, Wexford H...                       7   
Dovercourt Village, Dufferin                                            15   
Downsview Central                                                        3   
Downsview Northwest                                                      4   
Downsview West                                                           4   
Downsview, North Park, Upwood Park                                       4   
East Birchmount Park, Ionview, Kennedy Park                              5   
East Toronto                                                             3   
Emery, Humberlea                                                         1   
Fairview, Henry Farm, Oriole                                            71   
First Canadian Place, Underground city                                 100   
Flemingdon Park, Don Mills South                                        23   
Forest Hill North, Forest Hill West                                      4   
Glencairn                                                                4   
Guildwood, Morningside, West Hill                                        9   
Harbord, University of Toronto                                          31

#### Let's find out how many unique categories can be curated from all the returned venues


In [28]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 276 uniques categories.


#### Analyze each neighbourhood.

In [29]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0         Parkwoods                  0               0                  0   
1         Parkwoods                  0               0                  0   
2         Parkwoods                  0               0                  0   
3  Victoria Village                  0               0                  0   
4  Victoria Village                  0               0                  0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0     0    0               0                 0                 0   
1     0    0               0                 0                 0   
2     0    0               0                 0                 0   
3     0    0               0                 0                 0   
4     0    0               0                 0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Rental / Bike Share  Bike Shop  Bistro  \
0                   0                         0          0       0   
1                   0                         0          0       0   
2                   0                         0          0       0   
3                   0                         0          0       0   
4                   0                         0          0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0          

In [30]:
toronto_onehot.shape

(2125, 277)

#### Let's find out how many unique categories can be curated from all the returned venues


In [31]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                             Adelaide, King, Richmond           0.000000   
1                                            Agincourt           0.000000   
2    Agincourt North, L'Amoreaux East, Milliken, St...           0.000000   
3    Albion Gardens, Beaumond Heights, Humbergate, ...           0.000000   
4                               Alderwood, Long Branch           0.000000   
5      Bathurst Manor, Downsview North, Wilson Heights           0.000000   
6                                      Bayview Village           0.000000   
7                    Bedford Park, Lawrence Manor East           0.000000   
8                                          Berczy Park           0.000000   
9                          Birch Cliff, Cliffside West           0.000000   
10   Bloordale Gardens, Eringate, Markland Wood, Ol...           0.000000   
11        Brockton, Exhibition Place, Parkdale Village           0.000000   
12   Business Reply Mail Processing Centre 969 Eastern           0.000000   
13                         CFB Toronto, Downsview East           0.000000   
14   CN Tower, Bathurst Quay, Island airport, Harbo...           0.000000   
15                         Cabbagetown, St. James Town           0.000000   
16                                 Caledonia-Fairbanks           0.000000   
17               Canada Post Gateway Processing Centre           0.000000   
18                                           Cedarbrae           0.000000   
19                                  Central Bay Street           0.000000   
20           Chinatown, Grange Park, Kensington Market           0.000000   
21                                            Christie           0.000000   
22                                Church and Wellesley           0.000000   
23                     Clairlea, Golden Mile, Oakridge           0.000000   
24             Clarks Corners, Sullivan, Tam O'Shanter           0.000000   
25     Cliffcrest, Cliffside, Scarborough Village West           0.000000   
26   Cloverdale Mall, Islington, Martin Grove, Prin...           0.000000   
27                      Commerce Court, Victoria Hotel           0.000000   
28                                          Davisville           0.000000   
29                                    Davisville North           0.000000   
30   Deer Park, Forest Hill SE, Rathnelly, South Hi...           0.000000   
31      Del Ray, Keelesdale, Mount Dennis, Silverthorn           0.000000   
32            Design Exchange, Toronto Dominion Centre           0.000000   
33                                     Don Mills North           0.000000   
34   Dorset Park, Scarborough Town Centre, Wexford ...           0.000000   
35                        Dovercourt Village, Dufferin           0.000000   
36                                   Downsview Central           0.000000   
37                                 Downsview Northwest           0.000000   
38                                      Downsview West           0.000000   
39                  Downsview, North Park, Upwood Park           0.000000   
40         East Birchmount Park, Ionview, Kennedy Park           0.000000   
41                                        East Toronto           0.000000   
42                                    Emery, Humberlea           0.000000   
43                        Fairview, Henry Farm, Oriole           0.028169   
44              First Canadian Place, Underground city           0.000000   
45                    Flemingdon Park, Don Mills South           0.000000   
46                 Forest Hill North, Forest Hill West           0.000000   
47                                           Glencairn           0.000000   
48                   Guildwood, Morningside, West Hill           0.000000   
49                      Harbord, University of Toronto           0.000000   
50   Harbourfront East, Toronto Islands, Union Station           0.000000   
51                           Harbourf

In [32]:
toronto_grouped.shape

(101, 277)

#### Let's find out how many unique categories can be curated from all the returned venues


In [33]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.09
1             Café  0.05
2       Restaurant  0.04
3  Thai Restaurant  0.04
4              Gym  0.03


----Agincourt----
                       venue  freq
0                     Lounge  0.25
1             Breakfast Spot  0.25
2  Latin American Restaurant  0.25
3               Skating Rink  0.25
4          Accessories Store  0.00


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                venue  freq
0          Playground  0.33
1                Park  0.33
2        Intersection  0.33
3  Miscellaneous Shop  0.00
4               Motel  0.00


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                  venue  freq
0           Pizza Place  0.22
1         Grocery Store  0.22
2  Fast Food Restaurant  0.11
3              Pharmacy  0.11
4   Fried Chicken Joint  0.11


----Alderwood, Long Branch----
            venue  

#### Let's put that into a _pandas_ dataframe


In [34]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [35]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2  Agincourt North, L'Amoreaux East, Milliken, St...          Intersection   
3  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                       Café            Restaurant       Thai Restaurant   
1  Latin American Restaurant          Skating Rink        Breakfast Spot   
2                 Playground                  Park        Discount Store   
3                Pizza Place   Fried Chicken Joint        Sandwich Place   
4                        Gym        Sandwich Place              Pharmacy   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store         Deli / Bodega                   Gym   
1               Dog Run          Dessert Shop    Dim Sum Restaurant   
2          Dance Studio         Deli / Bodega      Department Store   
3            Beer Store  Fast Food Restaurant              Pharmacy   
4                   Pub           Coffee Shop          Airport Gate   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0                 Hotel                   Steakhouse           Concert Hall  
1                 Diner               Discount Store    Distribution Center  
2          Dessert Shop           Dim Sum Restaurant                  Diner  
3        General Travel                    Drugstore       Doner Restaurant  
4           Curling Ice  Eastern European Restaurant    Dumpling Restaurant

### Cluster Toronto Neighbourhoods

Run _k_-means to cluster the neighborhood into 5 clusters.


In [36]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 4, 0, 0, 0, 2, 2, 2, 2])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.


In [37]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged# check the last columns!

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
7        M9A         Etobicoke   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
52       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
58       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
64       M6J      West Toronto   
66       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
72       M4K      East Toronto   
74       M5K  Downtown Toronto   
76       M6K      West Toronto   
79       M1L       Scarborough   
82       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
87       M5L  Downtown Toronto   
89       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
96       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
102      M6M              York   
106      M9M        North York   
108      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
116      M9N              York   
117      M1P       Scarborough   
120      M2P        North York   
121      M4P   Central Toronto   
122      M5P   Central Toronto   
124      M6P      West Toronto   
126      M9P         Etobicoke   
127      M1R       Scarborough   
129      M2R        North York   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
188      M1X       Scarborough   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  \
0                                            Parkwoods  43.753259 -79.329656   
1                                     Victoria Village  43.725882 -79.315572   
2                            Harbourfront, Regent Park  43.654260 -79.360636   
4                     Lawrence Heights, Lawrence Manor  43.718518 -79.464763   
6                                         Queen's Park  43.662301 -

In [39]:
toronto_merged.dropna(subset=["Cluster Labels"], axis=0, inplace=True)
toronto_merged

Postcode           Borough  \
0        M3A        North York   
1        M4A        North York   
2        M5A  Downtown Toronto   
4        M6A        North York   
6        M7A      Queen's Park   
8        M1B       Scarborough   
10       M3B        North York   
11       M4B         East York   
13       M5B  Downtown Toronto   
15       M6B        North York   
16       M9B         Etobicoke   
21       M1C       Scarborough   
24       M3C        North York   
26       M4C         East York   
27       M5C  Downtown Toronto   
28       M6C              York   
29       M9C         Etobicoke   
33       M1E       Scarborough   
36       M4E      East Toronto   
37       M5E  Downtown Toronto   
38       M6E              York   
39       M1G       Scarborough   
40       M4G         East York   
41       M5G  Downtown Toronto   
42       M6G  Downtown Toronto   
43       M1H       Scarborough   
44       M2H        North York   
45       M3H        North York   
48       M4H         East York   
49       M5H  Downtown Toronto   
52       M6H      West Toronto   
54       M1J       Scarborough   
55       M2J        North York   
58       M3J        North York   
60       M4J         East York   
61       M5J  Downtown Toronto   
64       M6J      West Toronto   
66       M1K       Scarborough   
69       M2K        North York   
70       M3K        North York   
72       M4K      East Toronto   
74       M5K  Downtown Toronto   
76       M6K      West Toronto   
79       M1L       Scarborough   
82       M2L        North York   
84       M3L        North York   
85       M4L      East Toronto   
87       M5L  Downtown Toronto   
89       M6L        North York   
92       M9L        North York   
93       M1M       Scarborough   
96       M2M        North York   
98       M3M        North York   
99       M4M      East Toronto   
100      M5M        North York   
102      M6M              York   
106      M9M        North York   
108      M1N       Scarborough   
110      M2N        North York   
111      M3N        North York   
112      M4N   Central Toronto   
113      M5N   Central Toronto   
114      M6N              York   
116      M9N              York   
117      M1P       Scarborough   
120      M2P        North York   
121      M4P   Central Toronto   
122      M5P   Central Toronto   
124      M6P      West Toronto   
126      M9P         Etobicoke   
127      M1R       Scarborough   
129      M2R        North York   
130      M4R   Central Toronto   
131      M5R   Central Toronto   
134      M6R      West Toronto   
136      M7R       Mississauga   
137      M9R         Etobicoke   
141      M1S       Scarborough   
142      M4S   Central Toronto   
143      M5S  Downtown Toronto   
145      M6S      West Toronto   
147      M1T       Scarborough   
150      M4T   Central Toronto   
152      M5T  Downtown Toronto   
155      M1V       Scarborough   
159      M4V   Central Toronto   
164      M5V  Downtown Toronto   
171      M8V         Etobicoke   
174      M9V         Etobicoke   
182      M1W       Scarborough   
183      M4W  Downtown Toronto   
184      M5W  Downtown Toronto   
185      M8W         Etobicoke   
187      M9W         Etobicoke   
189      M4X  Downtown Toronto   
191      M5X  Downtown Toronto   
193      M8X         Etobicoke   
196      M4Y  Downtown Toronto   
197      M7Y      East Toronto   
198      M8Y         Etobicoke   
206      M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  \
0                                            Parkwoods  43.753259 -79.329656   
1                                     Victoria Village  43.725882 -79.315572   
2                            Harbourfront, Regent Park  43.654260 -79.360636   
4                     Lawrence Heights, Lawrence Manor  43.718518 -79.464763   
6                                         Queen's Park  43.662301 -79.389494   
8                                       Rouge, Malvern 

Finally, let's visualize the resulting clusters


In [40]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Examine Clusters

#### Cluster 1


In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             0.0          Intersection   
6        Queen's Park             0.0           Coffee Shop   
11          East York             0.0           Pizza Place   
21        Scarborough             0.0                   Bar   
29          Etobicoke             0.0          Liquor Store   
39        Scarborough             0.0           Coffee Shop   
41   Downtown Toronto             0.0           Coffee Shop   
45         North York             0.0           Coffee Shop   
58         North York             0.0  Caribbean Restaurant   
66        Scarborough             0.0    Chinese Restaurant   
92         North York             0.0           Pizza Place   
102              York             0.0           Coffee Shop   
114              York             0.0           Pizza Place   
126         Etobicoke             0.0           Pizza Place   
129        North York             0.0           Pizza Place   
131   Central Toronto             0.0                  Café   
136       Mississauga             0.0           Coffee Shop   
147       Scarborough             0.0           Pizza Place   
159   Central Toronto             0.0           Coffee Shop   
174         Etobicoke             0.0         Grocery Store   
182       Scarborough             0.0  Fast Food Restaurant   
185         Etobicoke             0.0           Pizza Place   
187         Etobicoke             0.0                   Bar   

    2nd Most Common Venue   3rd Most Common Venue      4th Most Common Venue  \
1             Pizza Place             Coffee Shop      Portuguese Restaurant   
6        Sushi Restaurant             Yoga Studio                      Diner   
11              Pet Store               Gastropub                       Café   
21            Yoga Studio                 Dog Run               Dessert Shop   
29             Beer Store                Pharmacy                Pizza Place   
39               Pharmacy            Soccer Field      Korean BBQ Restaurant   
41     Italian Restaurant                    Café             Sandwich Place   
45                   Bank                Pharmacy        Fried Chicken Joint   
58         Massage Studio             Coffee Shop         Falafel Restaurant   
66       Department Store             Coffee Shop             Discount Store   
92           Intersection  Furniture / Home Store                Curling Ice   
102    Turkish Restaurant          Sandwich Place                        Bar   
114        Breakfast Spot    Caribbean Restaurant                    Brewery   
126           Coffee Shop          Discount Store             Sandwich Place   
129               Butcher          Discount Store                Supermarket   
131        Sandwich Place             Coffee Shop                   Pharmacy   
136                 Hotel                     Gym  Middle Eastern Restaurant   
147  Fast Food Restaurant      Chinese Restaurant               Noodle House   
159    Light Rail Station     American Restaurant                 Bagel Shop   
174           Pizza Place     Fried Chicken Joint             Sandwich Place   
182           Pizza Place                    Bank                Coffee Shop   
185                   Gym          Sandwich Place                   Pharmacy   
187             Drugstore     Rental Car Location                 Truck Stop   

           5th Most Common Venue 6th Most Common Venue  \
1                   Hockey Arena        Discount Store   
6                 Sandwich Place           Salad Place   
11                          Bank    Athletics & Sports   
21            Dim Sum Restaurant                 Diner   
29                   Coffee Shop     Convenience Store   
39   Eastern European Restaurant   Dumpling Restaurant   
41                   Salad Place       Bubble Tea Shop   
45                 Shopping Mall                 Diner   
58                           Bar         Metro Station   


#### Cluster 2


In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
106  North York             1.0        Baseball Field           Yoga Studio   
198   Etobicoke             1.0        Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
106          Dessert Shop    Dim Sum Restaurant                 Diner   
198          Dessert Shop    Dim Sum Restaurant                 Diner   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
106        Discount Store   Distribution Center               Dog Run   
198        Discount Store   Distribution Center               Dog Run   

    9th Most Common Venue 10th Most Common Venue  
106      Doner Restaurant          Deli / Bodega  
198      Doner Restaurant          Deli / Bodega

#### Cluster 3


In [43]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
2    Downtown Toronto             2.0                 Coffee Shop   
4          North York             2.0              Clothing Store   
8         Scarborough             2.0        Fast Food Restaurant   
10         North York             2.0                         Gym   
13   Downtown Toronto             2.0                 Coffee Shop   
15         North York             2.0                      Bakery   
16          Etobicoke             2.0                  Print Shop   
24         North York             2.0                         Gym   
26          East York             2.0                Intersection   
27   Downtown Toronto             2.0                 Coffee Shop   
28               York             2.0                        Park   
33        Scarborough             2.0           Electronics Store   
36       East Toronto             2.0                Neighborhood   
37   Downtown Toronto             2.0                 Coffee Shop   
40          East York             2.0                 Coffee Shop   
42   Downtown Toronto             2.0               Grocery Store   
43        Scarborough             2.0         Fried Chicken Joint   
44         North York             2.0                 Golf Course   
48          East York             2.0           Indian Restaurant   
49   Downtown Toronto             2.0                 Coffee Shop   
52       West Toronto             2.0                    Pharmacy   
54        Scarborough             2.0                  Playground   
55         North York             2.0              Clothing Store   
61   Downtown Toronto             2.0                 Coffee Shop   
64       West Toronto             2.0                         Bar   
69         North York             2.0          Chinese Restaurant   
72       East Toronto             2.0            Greek Restaurant   
74   Downtown Toronto             2.0                 Coffee Shop   
76       West Toronto             2.0                        Café   
79        Scarborough             2.0                    Bus Line   
84         North York             2.0               Shopping Mall   
85       East Toronto             2.0        Fast Food Restaurant   
87   Downtown Toronto             2.0                 Coffee Shop   
89         North York             2.0  Construction & Landscaping   
93        Scarborough             2.0                       Motel   
96         North York             2.0                   Piano Bar   
98         North York             2.0                  Food Truck   
99       East Toronto             2.0                 Coffee Shop   
100        North York             2.0              Sandwich Place   
108       Scarborough             2.0             College Stadium   
110        North York             2.0                 Coffee Shop   
111        North York             2.0                Liquor Store   
112   Central Toronto             2.0            Business Service   
113   Central Toronto             2.0                      Garden   
117       Scarborough             2.0           Indian Restaurant   
121   Central Toronto             2.0                       Hotel   
122   Central Toronto             2.0               Jewelry Store   
124      West Toronto             2.0                        Café   
127       Scarborough             2.0   Middle Eastern Restaurant   
130   Central Toronto             2.0              Clothing Store   
134      West Toronto             2.0                   Gift Shop   
141       Scarborough             2.0                      Lounge   
142   Central Toronto             2.0                Dessert Shop   
143  Downtown Toronto             2.0                        Café   
145      West Toronto             2.0                 Coffee Shop   
150   Central Toronto             2.0                 Summer Camp   
152  Downtown Toronto             2.0                        Café   
164  Downtow

#### Cluster 4


In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
82  North York             3.0             Cafeteria           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
82   Distribution Center      Department Store          Dessert Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
82    Dim Sum Restaurant                 Diner        Discount Store   

   9th Most Common Venue 10th Most Common Venue  
82               Dog Run           Dance Studio

#### Cluster 5


In [45]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0          North York             4.0                  Park   
38               York             4.0                  Park   
60          East York             4.0                  Park   
70         North York             4.0               Airport   
116              York             4.0                  Park   
120        North York             4.0                  Park   
137         Etobicoke             4.0        Sandwich Place   
155       Scarborough             4.0          Intersection   
183  Downtown Toronto             4.0                  Park   
193         Etobicoke             4.0                  Park   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0    Fast Food Restaurant     Food & Drink Shop           Yoga Studio   
38          Women's Store                  Pool   Distribution Center   
60      Convenience Store           Yoga Studio        Discount Store   
70                   Park           Yoga Studio               Dog Run   
116     Convenience Store         Jewelry Store           Yoga Studio   
120     Convenience Store           Yoga Studio        Discount Store   
137                  Park           Yoga Studio        Discount Store   
155            Playground                  Park        Discount Store   
183            Playground                 Trail           Yoga Studio   
193                 River                  Pool           Yoga Studio   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Discount Store         Deli / Bodega      Department Store   
38          Deli / Bodega      Department Store          Dessert Shop   
60          Deli / Bodega      Department Store          Dessert Shop   
70       Department Store          Dessert Shop    Dim Sum Restaurant   
116          Dessert Shop    Dim Sum Restaurant                 Diner   
120         Deli / Bodega      Department Store          Dessert Shop   
137         Deli / Bodega      Department Store          Dessert Shop   
155          Dance Studio         Deli / Bodega      Department Store   
183                 Diner          Dance Studio         Deli / Bodega   
193                 Diner          Dance Studio         Deli / Bodega   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Dessert Shop    Dim Sum Restaurant                  Diner  
38     Dim Sum Restaurant                 Diner         Discount Store  
60     Dim Sum Restaurant                 Diner    Distribution Center  
70                  Diner        Discount Store    Distribution Center  
116        Discount Store   Distribution Center                Dog Run  
120    Dim Sum Restaurant                 Diner    Distribution Center  
137    Dim Sum Restaurant                 Diner    Distribution Center  
155          Dessert Shop    Dim Sum Restaurant                  Diner  
183      Department Store          Dessert Shop     Dim Sum Restaurant  
193      Department Store          Dessert Shop     Dim Sum Restaurant

After analyzing above clusters, It appears tha cluster 1 has most food stores among all the clusters. Cluster 3 and Cluster 5 also have many restaurants and other variety of stores.